win32ss/drivers/miniport/bochs/bochsmp.c

```diff
@@ -0,0 +1,744 @@
+          BochsWriteDispIAndCheck(DeviceExtension, VBE_DISPI_INDEX_BPP, 32);
+    /* Always enable screen, even if display settings change failed */
+    BochsWriteDispI(DeviceExtension, VBE_DISPI_INDEX_ENABLE, VBE_DISPI_LFB_ENABLED | VBE_DISPI_ENABLED);
+    if (!Ret)
+    {
+        VideoDebugPrint((Error, "Bochs: failed to change mode\n"));
+        return FALSE;
+    }
+
+    /* Enable VGA (QEMU secondary-vga disables it by default) */
+    if (!DeviceExtension->IoPorts.RangeInIoSpace)
+    {
+        /* Discard AR flip-flip */
+        (VOID)VideoPortReadRegisterUshort((PUSHORT)(DeviceExtension->IoPorts.Mapped + 0x41A));
+        /* Enable display */
+        VideoPortWriteRegisterUshort((PUSHORT)(DeviceExtension->IoPorts.Mapped + 0x400), 0x20);
+    }
+
+    DeviceExtension->CurrentMode = (USHORT)ModeRequested;
+    StatusBlock->Status = NO_ERROR;
+
+    VideoDebugPrint((Info, "Bochs:BochsSetCurrentMode Exit Mode:%d\n", ModeRequested));
+    return TRUE;
+}
+
+static BOOLEAN
+BochsQueryCurrentMode(
+    _In_ PBOCHS_DEVICE_EXTENSION DeviceExtension,
+    _Out_ PVIDEO_MODE_INFORMATION VideoModeInfo,
+    _Out_ PSTATUS_BLOCK StatusBlock)
+{
+    PBOCHS_SIZE AvailableModeInfo;
+
+    if (DeviceExtension->CurrentMode > DeviceExtension->AvailableModeCount)
+    {
+        StatusBlock->Status = ERROR_INVALID_PARAMETER;
+        return FALSE;
+    }
+
+    AvailableModeInfo = &DeviceExtension->AvailableModeInfo[DeviceExtension->CurrentMode];
+    VideoPortZeroMemory(VideoModeInfo, sizeof(*VideoModeInfo));
+    BochsGetModeInfo(AvailableModeInfo, VideoModeInfo, DeviceExtension->CurrentMode);
+
+    StatusBlock->Information = sizeof(*VideoModeInfo);
+    StatusBlock->Status = NO_ERROR;
+    return TRUE;
+}
+
+static BOOLEAN
+BochsResetDevice(
+    _In_ PBOCHS_DEVICE_EXTENSION DeviceExtension,
+    _Out_ PSTATUS_BLOCK StatusBlock)
+{
+    VideoDebugPrint((Info, "Bochs:BochsResetDevice Entry\n"));
+
+    StatusBlock->Status = NO_ERROR;
+
+    VideoDebugPrint((Info, "Bochs:BochsResetDevice Exit\n"));
+    return TRUE;
+}
+
+static BOOLEAN
+BochsGetChildState(
+    _In_ PBOCHS_DEVICE_EXTENSION DeviceExtension,
+    _Out_ PULONG pChildState,
+    _Out_ PSTATUS_BLOCK StatusBlock)
+{
+    *pChildState = VIDEO_CHILD_ACTIVE;
+
+    StatusBlock->Information = sizeof(*pChildState);
+    StatusBlock->Status = NO_ERROR;
+    return TRUE;
+}
+
+VP_STATUS NTAPI
+BochsFindAdapter(
+    _In_ PVOID HwDeviceExtension,
+    _In_ PVOID HwContext,
+    _In_ PWSTR ArgumentString,
+    _In_ PVIDEO_PORT_CONFIG_INFO ConfigInfo,
+    _In_ PUCHAR Again)
+{
+    if (ConfigInfo->Length < sizeof(*ConfigInfo))
+        return ERROR_INVALID_PARAMETER;
+
+    return NO_ERROR;
+}
+
+BOOLEAN NTAPI
+BochsInitialize(
+    _In_ PVOID HwDeviceExtension)
+{
+    ULONG PotentialModeCount = 0;
+    PBOCHS_DEVICE_EXTENSION DeviceExtension = HwDeviceExtension;
+
+    VideoDebugPrint((Info, "Bochs: BochsInitialize\n"));
+
+    if (!BochsGetControllerInfo(DeviceExtension))
```

Comment:
This is typically done within a call to `FindAdapter` routine.

Reply:
Moved code which doesn't access adapter from `BochsInitialize` to `BochsFindAdapter`